Add client span to SignalR .NET client

JamesNK opened the pull request:
Addresses https://github.com/dotnet/aspnetcore/issues/51557. Specifically, the client activity source item.

Follow up to https://github.com/dotnet/aspnetcore/pull/55439 and builds on top of https://github.com/dotnet/aspnetcore/pull/57049.

Changes:

* Invocations by the SignalR client now start a local activity.
* Activities are started for regular invocations and streaming methods.
* Activity starts with the invocation and ends when the client receives a completion message.

Not planned in this PR:

* Tracing in other clients.

Nothing is blocking additional work in the future in the future.

JamesNK committed:
Add client span to SignalR .NET client

JamesNK commented:
@BrennanConroy A problem I found when implementing this is the SignalR client doesn't know the name of the hub on the server. The client just specifies a URL and then sends messages to a method at that URL.

Because the hub name isn't available, as a substitute I replaced it with the URL path, e.g. `builder.WithUrl("https://localhost:5001/chat")` will result in the URL path `chat` being recorded as the service name.

Is there a better option? If not, should the server record the URL path as the OTEL service name instead of the .NET type name? That will keep consistency between what the client and server record.

JamesNK commented:
@tarekgh @noahfalk Do you know how the Activity name interacts with OTEL? Activity has OperationName and DisplayName. Right now, `Activity.OperationName` is set to the [required OTEL span format](https://github.com/open-telemetry/semantic-conventions/blob/aea69f203fd442c1dbdbb1479875f4de58d184d2/docs/rpc/rpc-spans.md#span-name). Is this the right thing to do?

It seems like opentelemetry-dotnet uses `Activity.DisplayName` (which defaults to OperationName if not set). And the [docs for Activity.OperationName](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.operationname?view=net-8.0) say its value should be a constant.

Would it be more correct to do this?

```cs
var activity = activitySource.CreateActivity("Microsoft.AspNetCore.SignalR.Client.InvocationOut", ...);
activity.DisplayName = "{serviceName}/{serviceMethod}";
```

BrennanConroy commented on src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:
```diff
@@ -1031,13 +1048,52 @@ private async Task CommonStreaming(ConnectionState connectionState, string strea
+
+        var activityName = !string.IsNullOrEmpty(serviceName)
+            ? $"{serviceName}/{methodName}"
+            : methodName;
```
Should we check if the ActivitySource (or Critical logging) is being listened to before allocating strings?

BrennanConroy commented on src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:
```diff
@@ -1031,13 +1048,52 @@ private async Task CommonStreaming(ConnectionState connectionState, string strea
+    private Activity? StartActivity(ConnectionContext connection, string methodName)
+    {
+        var connectionUri = (connection.RemoteEndPoint is UriEndPoint e) ? e.Uri : null;
+        var serviceName = connectionUri?.AbsolutePath.TrimStart('/').TrimEnd('/');
```
This info is constant for the lifetime of an active connection, should we just set it in `StartAsyncCore`?

JamesNK replied:
Yes, that can be done.

I asked some questions about service name [here](https://github.com/dotnet/aspnetcore/pull/57101#issuecomment-2260489868). What do you think?

BrennanConroy commented on src/SignalR/clients/csharp/Client.Core/src/Internal/InvocationRequest.cs:
```diff
@@ -13,15 +15,17 @@ namespace Microsoft.AspNetCore.SignalR.Client.Internal;
 internal abstract partial class InvocationRequest : IDisposable
 {
     private readonly CancellationTokenRegistration _cancellationTokenRegistration;
+    private int _isActivityStopping;
```
Did you find an issue with requests being completed/canceled multiple times?

Maybe if it was canceled by the user and completed by a response at the same time?

JamesNK replied:
`Cancel()` is always called from `Dispose()` regardless of whether `Complete()` or `Fail()` is already called. So that needed to be guarded against.

I wasn't sure whether these methods could happen simultaneously. All the other APIs called from them looked thread safe so I guessed probably and used Interlocked to make updating Activity thread safe.

A race between cancellation and complete/fail is the main scenario I thought of. I know that is possible in the gRPC client.

JamesNK replied:
> `Cancel()` is always called from `Dispose()` regardless of whether `Complete()` or `Fail()` is already called. So that needed to be guarded against.

By the way, this allocates an extra exception for streaming. `OperationCanceledException` is always allocated on invocation dispose, even if the stream was successfully completed.

https://github.com/dotnet/aspnetcore/blob/42e61729ecee15701ba55f4dbc6dc62d966a2a7f/src/SignalR/clients/csharp/Client.Core/src/Internal/InvocationRequest.cs#L124

BrennanConroy reviewed (commented):
Can we see a screenshot of these changes in a UI like the Aspire Dashboard?

`SendCoreAsync` doesn't have an activity being created for it. It's for "fire and forget" calls. Do we just create an activity, send the message, and stop the activity immediately?

There should be a few more tests:
Multiple different invokes to make sure there is a different activity every time.
Creating an activity in "user code" and seeing that it's the parent of the activity created in an invoke.
Making sure the activity stopped.

cijothomas commented:
> CC @cijothomas @CodeBlanch @reyang to answer the question in [#57101 (comment)](https://github.com/dotnet/aspnetcore/pull/57101#issuecomment-2260611442). What @JamesNK propose is matching the runtime guidelines as `OperationName` used for grouping and filtering while `DisplayName` used for the UI.

> Do you know how the Activity name interacts with OTEL? Activity has OperationName and DisplayName.

OTel has a single concept of Span Name, which is expected to be the low cardinality Spanname, also used for UI. The name *can* be changed after span creation. `Activity.DisplayName` is what maps to this. (Operationname exists as legacy on Activity, not used by OTel).

JamesNK commented:
@cijothomas Thanks!

JamesNK commented:
> Can we see a screenshot of these changes in a UI like the Aspire Dashboard?
> 
> `SendCoreAsync` doesn't have an activity being created for it. It's for "fire and forget" calls. Do we just create an activity, send the message, and stop the activity immediately?

Yes.

> Multiple different invokes to make sure there is a different activity every time.

This is done in `HubConnectionTests.InvokeAsync_SendTraceHeader`.

> Creating an activity in "user code" and seeing that it's the parent of the activity created in an invoke.

Done in `HubConnectionTests.InvokeAsync_SendTraceHeader` and `StreamAsyncCore_SendTraceHeader`

> Making sure the activity stopped.

I'll add some asserts for `Activity.IsStopped`.

BrennanConroy commented:
> > Multiple different invokes to make sure there is a different activity every time.
> 
> This is done in `HubConnectionTests.InvokeAsync_SendTraceHeader`.
> 
> > Creating an activity in "user code" and seeing that it's the parent of the activity created in an invoke.
> 
> Done in `HubConnectionTests.InvokeAsync_SendTraceHeader` and `StreamAsyncCore_SendTraceHeader`

Huh, maybe there is something weird happening here since it's merging into another PR, but those tests look like they're being deleted, which would be odd since they seem specific to this change.

JamesNK commented:
They're deleted from their original location and moved to `HubConnectionTests.Tracing.cs`.

BrennanConroy commented:
🙈 The file was collapsed and when I looked for HubConnectionTests.Tracing.cs I only noticed one of them but there are two files with that name.

JamesNK commented:
Updates:

* Client activity for SendAsync added.
* Changed OperationName to a constant. Will fix server activity in another PR.
* Changed client activity to include acquiring an active connection.

I don't think there are any outstanding changes. There is still this open question: https://github.com/dotnet/aspnetcore/pull/57101#issuecomment-2260489868 @BrennanConroy

JamesNK commented:
> Can we see a screenshot of these changes in a UI like the Aspire Dashboard?

![image](https://github.com/user-attachments/assets/560459fd-e339-43ac-81ae-abd0f6c3415f)

😎 😎 😎

JamesNK commented on src/SignalR/clients/csharp/Client.Core/src/Internal/InvocationRequest.cs:
```diff
@@ -82,7 +93,15 @@ public override void Complete(CompletionMessage completionMessage)
+                }
+
                 _channel.Writer.TryComplete(new InvalidOperationException("Server provided a result in a completion response to a streamed invocation."));
+                return;
```
@BrennanConroy Is adding this return correct? It seems like this method should exit if there is no result, and the return was missing.

JamesNK replied:
Not returning and calling fail after this will log the error but the user won't see it from the channel reader. Instead, they see the exception reporting that the stream incorrectly returned a response.

Is that what you want?

BrennanConroy replied:
Actually, I think the added return is good. You're not supposed to set both error and result (and the server won't), and like you point out, the channel will already be completed with an InvalidOperationException.

BrennanConroy commented on src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:
```diff
@@ -235,6 +239,10 @@ public HubConnection(IConnectionFactory connectionFactory, IHubProtocol protocol
 
+        // ActivitySource can be resolved from the service provider when unit testing.
+        _activitySource = (serviceProvider.GetService<SignalRClientActivitySource>() ?? SignalRClientActivitySource.Instance).ActivitySource;
+        _serviceName = (_endPoint is UriEndPoint e) ? e.Uri.AbsolutePath.Trim('/') : null;
```
I think this is fine, we don't know the class name on the client side and like you said elsewhere the server-side should have the route to match against already.

Do you know if other frameworks like gRPC have a similar problem?

JamesNK replied:
The client and server both know the service name with gRPC

BrennanConroy commented on src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:
```diff
@@ -746,6 +757,13 @@ async Task OnStreamCanceled(InvocationRequest irq)
             _state.ReleaseConnectionLock();
         }
 
+        // The streaming method can continue after control is returned to the caller.
```
InvokeAsync would have the same issue if you assigned the result task right?
```csharp
var invokeTask = connection.InvokeAsync<int>("Get");

// Activity.Current is now the invoke activity

await invokeTask;
```

JamesNK replied:
Maybe. I'll test.

JamesNK replied:
It's fine. I'm a little unsure how it does the right thing, but it's doing the right thing.

BrennanConroy commented on src/SignalR/clients/csharp/Http.Connections.Client/src/HttpConnection.cs:
```diff
@@ -148,6 +148,8 @@ public HttpConnection(HttpConnectionOptions httpConnectionOptions, ILoggerFactor
 
         _url = _httpConnectionOptions.Url;
 
+        RemoteEndPoint = new UriEndPoint(_url);
```
I wonder if this should be set after negotiate finishes, since negotiate can redirect to another endpoint, e.g. Azure SignalR Service

JamesNK replied:
Right now this actually isn’t used. The activity uses values from _endpoint

JamesNK replied:
How does negotiate change the url? Is there an example in tests of this happening?

The invocation activity is started before the connection is acquired. I could change logic to update server.address and server.port if the connection reports different value.

BrennanConroy replied:
https://github.com/dotnet/aspnetcore/blob/26ee437e46d8cdbea7113833942690c2ad2a28c8/src/SignalR/clients/csharp/Client/test/UnitTests/HttpConnectionTests.Negotiate.cs#L415

Spec that talks about it:
https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/docs/specs/TransportProtocols.md#all-versions

JamesNK replied:
Done. Client span's now report the negotiated URL.

[`a698667` (#57101)](https://github.com/dotnet/aspnetcore/pull/57101/commits/a698667a9478e6be87395eb4bc9987b4158e7c15)

BrennanConroy reviewed (commented):
I think this is good to go after my last questions/comments.

JamesNK commented:
> I think this is good to go after my last questions/comments.

This PR depends on the propagation PR being merged first

JamesNK commented on src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:
```diff
@@ -720,8 +728,11 @@ async Task OnStreamCanceled(InvocationRequest irq)
+        var activity = StartActivity(methodName);
+        var connectionState = await WaitForActiveConnectionWithActivityAsync(nameof(StreamAsChannelCoreAsync), activity, token: cancellationToken).ConfigureAwait(false);
+
+        Activity? previousActivity = Activity.Current;
```
I didn't move it when I moved the activity. It didn't break a test because it isn't needed. Removed.

JamesNK committed:
Fix merge

JamesNK committed:
Handle negotiate changing the connection URL

BrennanConroy commented on src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs:
```diff
@@ -1003,12 +1012,71 @@ private async Task CommonStreaming(ConnectionState connectionState, string strea
+        {
+            ConnectionState connectionState;
+            var connectionStateTask = _state.WaitForActiveConnectionAsync(sendingMethodName, token);
+            if (connectionStateTask.Status == TaskStatus.RanToCompletion)
```
Do we need to set any server tags if `connectionStateTask` throws. It throws when you're either not connected or the user canceled the passed in CTS. In either case it's not actually doing anything except creating and stopping an activity.

Plus, it'd clean up the code a bit.

JamesNK replied:
If there is a problem starting the connection then it would be useful to know the server address being called

BrennanConroy replied:
But the connection failing to start, or closing before the send calls, or not even calling start in the first place, is not reflected in SendAsync or InvokeAsync calls. All it knows is that there is no connection.

JamesNK replied:
With telemetry it's useful to view it from the perspective of the user. The mechanics of what happens internally that could cause the call to fail are invisible to them. They're focused on the end result: the call they want to make failed and there was an error. Having information about what they were trying to do - make a call to a hub with the specified `rpc.service`, `rpc.method`, `server.address`, `server.port` - is what they care about. Then the error reason would be in `error.type`.

If you want to make it clearer that the call failed because of the connection not being available, there is room to do that by providing good values to `error.type`. By default, it is the exception type name, but it can be customized in known scenarios. For example, if there is a problem with the connection, it could be something like: `negotiate-failed`, `hub-connection-not-started`, etc.

JamesNK replied:
Also, `server.address` and `server.port` are considered required attributes according to the spec: https://github.com/open-telemetry/semantic-conventions/blob/1e34b57b9f73b08b109cdc0e8841e857e5f5c205/docs/rpc/rpc-spans.md#client-attributes

BrennanConroy replied:
Ok thanks, I wanted to know some concrete reason we were including it.

BrennanConroy commented on src/SignalR/clients/csharp/Client/test/UnitTests/HubConnectionTests.Tracing.cs:
```diff
@@ -124,5 +127,91 @@ public async Task SendAnInvocationMessage_SendTraceHeaders()
+            TestConnection connection = null;
+            connection = new TestConnection(onStart: () =>
+            {
+                connection.RemoteEndPoint = new UriEndPoint(new Uri("http://example.net"));
```
nit: Should this be `"http://example.net:5050"` and then asserting `"server.port"`?

JamesNK committed:
Test port

JamesNK commented:
/azp run

JamesNK committed:
Fix test

JamesNK commented:
/azp run
